src/control/lib/telemetry/telemetry.go

```diff
@@ -0,0 +1,396 @@
+	MetricTypeTimestamp MetricType = C.D_TM_TIMESTAMP
+)
+
+type (
+	Metric interface {
+		Path() string
+		Name() string
+		Type() MetricType
+		ShortDesc() string
+		LongDesc() string
+		FloatValue() float64
+		String() string
+	}
+
+	StatsMetric interface {
+		Metric
+		FloatMin() float64
+		FloatMax() float64
+		FloatSum() float64
+		Mean() float64
+		StdDev() float64
+		SampleSize() uint64
+	}
+)
+
+type (
+	handle struct {
+		sync.RWMutex
+		idx   uint32
+		rank  *uint32
+		shmem *C.uint64_t
+		root  *C.struct_d_tm_node_t
+	}
+
+	metricBase struct {
+		handle *handle
+		node   *C.struct_d_tm_node_t
+
+		path      string
+		name      *string
+		shortDesc *string
+		longDesc  *string
+	}
+
+	statsMetric struct {
+		metricBase
+		stats C.struct_d_tm_stats_t
+	}
+
+	telemetryKey string
+)
+
+const (
+	handleKey telemetryKey = "handle"
+)
+
+func getHandle(ctx context.Context) (*handle, error) {
+	handle, ok := ctx.Value(handleKey).(*handle)
+	if !ok {
+		return nil, errors.New("no handle set on context")
+	}
+	return handle, nil
+}
+
+func findNode(hdl *handle, name string) (*C.struct_d_tm_node_t, error) {
+	if hdl == nil {
+		return nil, errors.New("nil handle")
+	}
+
+	node := C.d_tm_find_metric(hdl.shmem, C.CString(name))
+	if node == nil {
+		return nil, errors.Errorf("unable to find metric named %q", name)
+	}
+
+	return node, nil
+}
+
+func (mb *metricBase) Type() MetricType {
+	return MetricTypeUnknown
+}
+
+func (mb *metricBase) Path() string {
+	if mb == nil {
+		return "<nil>"
+	}
+	return mb.path
+}
+
+func (mb *metricBase) Name() string {
+	if mb == nil || mb.handle == nil || mb.node == nil {
+		return "<nil>"
+	}
+
+	if mb.name == nil {
+		name := C.GoString((*C.char)(C.d_tm_conv_ptr(mb.handle.shmem, unsafe.Pointer(mb.node.dtn_name))))
+		mb.name = &name
+	}
+
+	return *mb.name
+}
+
+func (mb *metricBase) fillMetadata() {
+	if mb == nil || mb.handle == nil || mb.handle.root == nil {
+		return
+	}
+
+	var shortDesc *C.char
+	var longDesc *C.char
+	res := C.d_tm_get_metadata(&shortDesc, &longDesc, mb.handle.shmem, mb.node, C.CString(mb.Name()))
+	if res == C.DER_SUCCESS {
+		short := C.GoString(shortDesc)
+		mb.shortDesc = &short
+		long := C.GoString(longDesc)
+		mb.longDesc = &long
+
+		C.free(unsafe.Pointer(shortDesc))
+		C.free(unsafe.Pointer(longDesc))
+	} else {
+		failed := "failed to retrieve metadata"
+		mb.shortDesc = &failed
+		mb.longDesc = &failed
+	}
+}
+
+func (mb *metricBase) ShortDesc() string {
+	if mb.shortDesc == nil {
+		mb.fillMetadata()
+	}
+
+	return *mb.shortDesc
+}
+
+func (mb *metricBase) LongDesc() string {
+	if mb.longDesc == nil {
+		mb.fillMetadata()
+	}
+
+	return *mb.longDesc
+}
+
+func (mb *metricBase) String() string {
+	r, w, err := os.Pipe()
+	if err != nil {
+		return err.Error()
+	}
+	defer r.Close()
+	defer w.Close()
+
+	f := C.fdopen(C.int(w.Fd()), C.CString("w"))
+	if f == nil {
+		return "fdopen() failed"
+	}
+
+	go func() {
+		C.d_tm_print_node(mb.handle.shmem, mb.node, C.int(0), f)
```

Comment:
The print function changed a bit in the latest update.  You can specify a format for verbose or CSV.  Maybe you will want to carry that option here, and make a Stringify for verbose and a Stringify for csv so there will be options for printing these through DMG or other utilities later.

Reply:
I think the default will be sufficient for the basic stringer implementation. The general practice for stringers is just to produce a human-comprehensible representation of the type, not to implement "business logic".